pkg/net/bitswap.go

```diff
@@ -151,3 +163,159 @@ func (c BitswapClient) Retrieve(
 		return task.NewErrorRetrievalResultWithErrorResolution(task.RetrievalFailure, err), nil
 	}
 }
+
+// Starts with the root CID, then fetches a random CID from the children and grandchildren nodes, until it reaches `traverseDepth`
+// Note: the root CID is considered depth `0`, so passing `traverseDepth=0` will only fetch the root CID
+// Returns a `SuccessfulRetrievalResult` if *all* retrievals were successful, `ErrorRetrievalResult` if any failed
+func (c BitswapClient) SpadeTraversal(parent context.Context, target peer.AddrInfo, startingCid cid.Cid, traverseDepth uint) (*task.RetrievalResult, error) {
+	logger := logging.Logger("bitswap_client_spade").With("cid", startingCid).With("target", target)
+	cidToRetrieve := startingCid
+
+	// Initialize hosts and clients required to do all the retrieval tests
+	network := bsnet.NewFromIpfsHost(c.host, SingleContentRouter{
+		AddrInfo: target,
+	})
+	bswap := bsclient.New(parent, network, blockstore.NewBlockstore(datastore.NewMapDatastore()))
+
+	defer bswap.Close()
+	defer network.Stop()
+
+	startTime := time.Now()
+
+	// support structures such as: https://github.com/filecoin-project/go-dagaggregator-unixfs#grouping-unixfs-structure
+	i := uint(0)
+	for {
+		// Retrieval
+		logger.Infof("retrieving %s\n", cidToRetrieve.String())
+		blk, err := c.RetrieveBlock(parent, target, network, bswap, cidToRetrieve)
+		if err != nil {
+			return task.NewErrorRetrievalResultWithErrorResolution(task.RetrievalFailure, err), nil
+		}
+
+		if i == traverseDepth {
+			var size = int64(len(blk.RawData()))
+			elapsed := time.Since(startTime)
+			logger.With("size", size).With("elapsed", elapsed).Info("Retrieved block")
+
+			// we've reached the requested depth of the tree
+			return task.NewSuccessfulRetrievalResult(elapsed, size, elapsed), nil
+		}
+
+		// if not at bottom of the tree, keep going down the links until we reach it
+		links, err := FindLinks(parent, blk)
+		if err != nil {
+			return task.NewErrorRetrievalResultWithErrorResolution(task.CannotDecodeLinks, err), nil
+		}
+
+		logger.Debugf("retrieved %s which has %d links\n", cidToRetrieve.String(), len(links))
+
+		nextIndex := 0
+		rand.Seed(time.Now().UnixNano())
+		if i == 0 {
+			// Special logic for when we are at the root node
+
+			if len(links) == 0 {
+				return task.NewErrorRetrievalResult(task.CannotTraverse, errors.Errorf("root node contains no links, cannot traverse any further")), nil
+			}
+			if len(links) == 1 {
+				// Generally this should not happen as the root node should contain at least one AggregateManifest and other links
+				// however, there may be a different construction in the future where this is still valid, so we will attempt to retrieve
+				logger.Info("starting node contains only 1 link - will still attempt retrieval test")
+				nextIndex = 0
+			} else {
+				// To be safe, never grab the first link off the root as it may refer to the AggregateManifest
+				nextIndex = 1 + rand.Intn(len(links)-1)
+			}
```

Comment:
See my comment about the `daggregator` thing being on its way out. Encoding all of this is a mistake.

Reply:
@ribasushi  I understand that, but I'm not sure if there's a better way to do it: 

To support the `dagaggregator` case, we essentially need to _skip link at index [0] of the root node_ as it _may_ be an AggregateManifest, which is not traversable. With this logic here, we skip index [0] (unless there's only one link) meaning that we'll never mistakenly hit the AggregateManifest. This does introduce a trade-off in the case of a non-dagaggregator retrieval test, as we're potentially not testing one subtree off the root. 

The only other ways I can think of doing this is: 
a) Specifically encode the fact that we're doing a `daggregator` retrieval test in `Task` itself, and have a switch statement with a separate logic flow for it. This way, we ring-fence it and can easily support any future structures by being explicit about it (i.e, add a `TestStrategy` variable to all Spade retrieval test tasks)

b) Remove this part of the code that skips `links[0]`, but then add some retry logic to sample a different subtree if it does happen to grab the AggregateManifest. This seems to be more complex to me and will result in a lot of retries.

I'd lean towards option a) if we want to be more explicit about it, but I could be missing some more context about how the future of things will look that might point towards a better strategy?

Reply:
as discussed, re-worked `traverse_depth` to `max_traverse_depth`, we now no longer fail if we find a node without links before getting to the specified depth.